aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -875,38 +912,22 @@ inline ScalarType get_dtype_from_result(Tensor& result, optional<ScalarType> dty
   }
 }
 
-inline ScalarType get_dtype_from_self(const Tensor& self, optional<ScalarType> dtype,
-                            bool promote_integers) {
-  if (dtype.has_value()) {
-    return dtype.value();
-  }
-  ScalarType src_type = self.scalar_type();
-  if (promote_integers && at::isIntegralType(src_type, /*includeBool=*/true)) {
-    return kLong;
-  }
-  return src_type;
-}
-
-Tensor& sum_out(const Tensor& self, IntArrayRef dim,
-                       bool keepdim, optional<ScalarType> opt_dtype, Tensor& result) {
-  ScalarType dtype = get_dtype_from_result(result, opt_dtype);
-  auto iter = make_reduction("sum", result, self, dim, keepdim, dtype);
+TORCH_IMPL_FUNC(sum_out)
+(const Tensor& self,
+ IntArrayRef dim,
+ bool keepdim,
+ optional<ScalarType> opt_dtype,
+ const Tensor& result) {
+  auto iter = meta::make_reduction_from_out_ty(self, result, dim, keepdim, result.scalar_type());
```

Comment:
Remind me again why `sum` doesn't use TensorIteratorBase directly?

Reply:
oh is this the thing where reductions are weird

Reply:
Yes! Basically, `TensorIteratorBase` expects the to-be-reduced dimensions to have size 1 and stride 0, so that it can reorder the dimensions. In the current state, I think there's no way to do that without changing the result's shape and properties.